File: examples/anip/anip_server/data/flights.py
"""Stub flight data — in-memory, no database."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class Flight:
    flight_number: str
    origin: str
    destination: str
    date: str
    departure_time: str
    arrival_time: str
    price: float
    currency: str = "USD"
    stops: int = 0


@dataclass
class Booking:
    booking_id: str
    flight: Flight
    passengers: int
    total_cost: float
    booked_by: str  # delegation chain subject
    on_behalf_of: str  # root principal


# Stub flight inventory
FLIGHTS: list[Flight] = [
    Flight("AA100", "SEA", "SFO", "2026-03-10", "08:00", "10:15", 420.00),
    Flight("UA205", "SEA", "SFO", "2026-03-10", "11:30", "13:45", 380.00),
    Flight("DL310", "SEA", "SFO", "2026-03-10", "14:00", "18:30", 280.00, stops=1),
    Flight("AA101", "SEA", "SFO", "2026-03-11", "08:00", "10:15", 310.00),
    Flight("UA450", "SEA", "LAX", "2026-03-10", "09:00", "11:30", 350.00),
    Flight("DL520", "SFO", "JFK", "2026-03-12", "06:00", "14:30", 580.00),
]

# In-memory booking store
_bookings: dict[str, Booking] = {}
_next_booking_id = 1


def search_flights(origin: str, destination: str, date: str) -> list[Flight]:
    return [
        f for f in FLIGHTS
        if f.origin == origin and f.destination == destination and f.date == date
    ]


def get_flight(flight_number: str, date: str) -> Flight | None:
    for f in FLIGHTS:
        if f.flight_number == flight_number and f.date == date:
            return f
    return None


def create_booking(
    flight: Flight,
    passengers: int,
    booked_by: str,
    on_behalf_of: str,
) -> Booking:
    global _next_booking_id
    booking_id = f"BK-{_next_booking_id:04d}"
    _next_booking_id += 1
    booking = Booking(
        booking_id=booking_id,
        flight=flight,
        passengers=passengers,
        total_cost=flight.price * passengers,
        booked_by=booked_by,
        on_behalf_of=on_behalf_of,
    )
    _bookings[booking_id] = booking
    return booking


def get_booking(booking_id: str) -> Booking | None:
    return _bookings.get(booking_id)

File: examples/anip-ts/src/domain/flights.ts
/**
 * Stub flight data — in-memory, no database.
 */

export interface Flight {
  flight_number: string;
  origin: string;
  destination: string;
  date: string;
  departure_time: string;
  arrival_time: string;
  price: number;
  currency: string;
  stops: number;
}

export interface Booking {
  booking_id: string;
  flight: Flight;
  passengers: number;
  total_cost: number;
  booked_by: string; // delegation chain subject
  on_behalf_of: string; // root principal
}

// Stub flight inventory
export const FLIGHTS: Flight[] = [
  {
    flight_number: "AA100",
    origin: "SEA",
    destination: "SFO",
    date: "2026-03-10",
    departure_time: "08:00",
    arrival_time: "10:15",
    price: 420.0,
    currency: "USD",
    stops: 0,
  },
  {
    flight_number: "UA205",
    origin: "SEA",
    destination: "SFO",
    date: "2026-03-10",
    departure_time: "11:30",
    arrival_time: "13:45",
    price: 380.0,
    currency: "USD",
    stops: 0,
  },
  {
    flight_number: "DL310",
    origin: "SEA",
    destination: "SFO",
    date: "2026-03-10",
    departure_time: "14:00",
    arrival_time: "18:30",
    price: 280.0,
    currency: "USD",
    stops: 1,
  },
  {
    flight_number: "AA101",
    origin: "SEA",
    destination: "SFO",
    date: "2026-03-11",
    departure_time: "08:00",
    arrival_time: "10:15",
    price: 310.0,
    currency: "USD",
    stops: 0,
  },
  {
    flight_number: "UA450",
    origin: "SEA",
    destination: "LAX",
    date: "2026-03-10",
    departure_time: "09:00",
    arrival_time: "11:30",
    price: 350.0,
    currency: "USD",
    stops: 0,
  },
  {
    flight_number: "DL520",
    origin: "SFO",
    destination: "JFK",
    date: "2026-03-12",
    departure_time: "06:00",
    arrival_time: "14:30",
    price: 580.0,
    currency: "USD",
    stops: 0,
  },
];

// In-memory booking store
const bookings: Map<string, Booking> = new Map();
let nextBookingId = 1;

export function searchFlights(
  origin: string,
  destination: string,
  date: string
): Flight[] {
  return FLIGHTS.filter(
    (f) => f.origin === origin && f.destination === destination && f.date === date
  );
}

export function getFlight(
  flightNumber: string,
  date: string
): Flight | null {
  for (const f of FLIGHTS) {
    if (f.flight_number === flightNumber && f.date === date) {
      return f;
    }
  }
  return null;
}

export function createBooking(
  flight: Flight,
  passengers: number,
  bookedBy: string,
  onBehalfOf: string
): Booking {
  const bookingId = `BK-${String(nextBookingId).padStart(4, "0")}`;
  nextBookingId += 1;
  const booking: Booking = {
    booking_id: bookingId,
    flight,
    passengers,
    total_cost: flight.price * passengers,
    booked_by: bookedBy,
    on_behalf_of: onBehalfOf,
  };
  bookings.set(bookingId, booking);
  return booking;
}

export function getBooking(bookingId: string): Booking | null {
  return bookings.get(bookingId) ?? null;
}
